dotnet/src/SemanticKernel.Core/Extensions/KernelSemanticFunctionExtensions.cs

```diff
@@ -285,4 +313,89 @@ public static class KernelSemanticFunctionExtensions
+    private const string BasicTemplateFactoryTypeName = "BasicPromptTemplateFactory";
+    private static bool s_promptTemplateFactoryInitialized = false;
+    private static Type? s_promptTemplateFactoryType = null;
+
+    /// <summary>
+    /// Create a default prompt template factory.
+    ///
+    /// This is a temporary solution to avoid breaking existing clients.
+    /// There will be a separate task to add support for registering instances of IPromptTemplateEngine and obsoleting the current approach.
+    ///
+    /// </summary>
+    /// <param name="kernel">Kernel instance</param>
+    /// <returns>Instance of <see cref="IPromptTemplateEngine"/>.</returns>
+    private static IPromptTemplateFactory CreateDefaultPromptTemplateFactory(IKernel kernel)
+    {
+#pragma warning disable CS0618 // Type or member is obsolete
+        if (kernel.PromptTemplateEngine is not null)
+        {
+            return new PromptTemplateFactory(kernel.PromptTemplateEngine);
+        }
+#pragma warning restore CS0618 // Type or member is obsolete
+
+        if (!s_promptTemplateFactoryInitialized)
+        {
+            s_promptTemplateFactoryType = GetPromptTemplateFactoryType();
+            s_promptTemplateFactoryInitialized = true;
+        }
+
+        if (s_promptTemplateFactoryType is not null)
+        {
+            var constructor = s_promptTemplateFactoryType.GetConstructor(new Type[] { typeof(ILoggerFactory) });
+            if (constructor is not null)
+            {
+#pragma warning disable CS8601 // Null logger factory is OK
+                var factory = (IPromptTemplateFactory)constructor.Invoke(new object[] { kernel.LoggerFactory });
+                if (factory is not null)
+                {
+                    return factory;
+                }
+#pragma warning restore CS8601
+            }
+        }
+
+        throw new SKException($"Unable to create default prompt template factory. Please provide an implementation of IPromptTemplateFactory or depend on {BasicTemplateFactoryAssemblyName}");
+    }
+
+    /// <summary>
+    /// Get the prompt template engine type if available
+    /// </summary>
+    /// <returns>The type for the prompt template engine if available</returns>
+    private static Type? GetPromptTemplateFactoryType()
+    {
+        try
+        {
+            var assembly = Assembly.Load(BasicTemplateFactoryAssemblyName);
+
+            return assembly.ExportedTypes.Single(type =>
+                type.Name.Equals(BasicTemplateFactoryTypeName, StringComparison.Ordinal) &&
+                type.GetInterface(nameof(IPromptTemplateFactory)) is not null);
+        }
+        catch (Exception ex) when (!ex.IsCriticalException())
+        {
+            return null;
+        }
+    }
+
+    #endregion
+}
+
+[Obsolete("IPromptTemplateEngine is being replaced with IPromptTemplateFactory. This will be removed in a future release.")]
+internal sealed class PromptTemplateFactory : IPromptTemplateFactory
```

Comment:
Move to another file or add this as a private inner class of the extension class.

Reply:
Done